### PR TITLE
Vertex tool should not select invisible features

### LINKED ISSUE #25259: Node Tool Selects Invisible Objects
Author Name: **Ryan Desmond** (@rddesmond)
Original Redmine Issue: [17362](https://issues.qgis.org/issues/17362)
Affected QGIS version: master
Redmine category:editing

---
The node tool, selection happens without regard to the visibility of the object.  In the simple case (procedure 1), you can simply set the layer to invisible and show that object selection is still possible.  In a more complex case (procedure 2), a portion of a layer is made invisible (with categorized style rules).

With the existing node tool, procedure one is a bit unnatural, because you can only select features on the currently selected layer.  The second procedure is a much more natural use case. 
 One or both of these cases may be OBE by the new node tool in 3.0.

## Procedure 1:

1. Create a polygon scratch layer
1. Edit the layer
1. Select add feature and insert an object
1. Select the "Node Tool"
1. Toggle layer visibility to hide it
1. Click where a node would be on the newly hidden object

## Procedure 2:

1. Create a polygon scratch layer
1. Edit the layer
1. Select add feature and insert two objects
1. Select "Open Attribute table"
1. Add a new column and categorize each of the objects uniquely (e.g. 1, 2, ..)
1. Select the "Node Tool"
1. In Properties / Style, set the style to be categorized by the new column and hit "classify"
1. Unselect one of the categories
1. Click where a node would be on the newly hidden object

## Expected

Visible objects are selectable, but invisible objects are not.

## Actual

In both cases, the objects which aren't visible are selected to edit.

### PR DESCRIPTION
**Pull request by uclaros**
## Description
This is how it's done in `QgsMapToolSelectUtils` for `QgsMapToolSelect`.
Is this a good way to handle this?

Fixes: #25259

### CODE REVIEW

**Comment by roya0045** [comment]
If this work this will be great! How does it deal with rule based features?

**Comment by uclaros** [comment]
Since it's using a copy of the layer's renderer, everything that's not rendered should not be selectable.

**Comment by roya0045** [comment]
I would think that a gif of the modification (since a test seems unlikely) would be beneficial since there are no way to see this in action other than to build ourselves. But the changes seems reasonable.

**Comment by saberraz** [comment]
I have tested your improvements but I am not 100% sure with this use-case:
Imagine, you are in topological edit mode, if a feature is hidden, and you try to use the node tool, it does not edit the hidden tool. I think for topological edit mode, this should work on all features and not only visible ones:

You can see in the screenshot below, the footpath is hidden (but visible in the 2nd map view). When I do node move for the main road, it works with all vertices. But, for footpath, it will leave it untouched.  I understand, your new logic to only apply to the visible features is more handy, but this might mess up a lot of data model, where they need to have the correct topology but only show the subset to user to edit:
edit

**Comment by uclaros** [comment]
> they need to have the correct topology but only show the subset to user to edit:

I hear your concern, but is it really safe to edit invisible data? Moving vertices with topological editing is kind of safe (you may introduce unwanted intersections), but deleting is not. What if the user deletes the seemingly unneeded vertex (a vertex in the middle of a straight segment) you moved? You also lose a segment of your footpath! With polygons it can get worse!
A possible solution would be to use an invisible style, eg.transparent, for the invisible features.

**Comment by roya0045** [comment]
I have to agree with ucarlos, what I dislike the most is having features from other layer act as editable feature when the layer is not displayed (though this may be fixed). With overlapping feature is can be nearly impossible to edit the desired feature.

**Comment by uclaros** [comment]
@saberraz I just tested on master and while the invisible vertices can be selected and moved, they are *not* moved by topological editing. This PR should affect only the drag-select functionality, ie. the invisible line's vertices should not be selected:

Peek 2020-04-16 23-29

**Comment by uclaros** [comment]
This PR is ready for review :)

File: src/app/vertextool/qgsvertextool.cpp [hunk 1]
```diff
@@ -545,17 +546,39 @@ void QgsVertexTool::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
...
+
       QgsRectangle layerRect = layerRubberBandGeometry.boundingBox();
       std::unique_ptr< QgsGeometryEngine > layerRubberBandEngine( QgsGeometry::createGeometryEngine( layerRubberBandGeometry.constGet() ) );
       layerRubberBandEngine->prepareGeometry();
 
+      QgsFeatureRequest request;
+      request.setFilterRect( layerRect );
+      request.setFlags( QgsFeatureRequest::ExactIntersect );
+      if ( r )
+        request.setSubsetOfAttributes( r->usedAttributes( context ), vlayer->fields() );
```
Comment (nyalldawson) [inline src/app/vertextool/qgsvertextool.cpp:566]:
```suggestion
    {
        request.setSubsetOfAttributes( r->usedAttributes( context ), vlayer->fields() );
        request.setFilterExpression( r->filter( vlayer->fields() ) );
     }
```

**Comment by nyalldawson** [comment]
We could also consider adding a warning message if invisible features were selected which would have otherwise been modified. (If we do that, you'll need to ignore the above suggestion!)

File: src/app/vertextool/qgsvertextool.cpp [hunk 2]
```diff
@@ -545,31 +548,71 @@ void QgsVertexTool::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
...
       while ( fi.nextFeature( f ) )
       {
         if ( mLockedFeature && mLockedFeature->featureId() != f.id() )
           continue;  // with locked feature we only allow selection of its vertices
 
+        context.expressionContext().setFeature( f );
+        // make sure to only use features that are visible
+        bool isFeatureInvisible = ( r && !r->willRenderFeature( f, context ) );
+        if ( isFeatureInvisible && invisibleVertexSelected )
+          continue;
```
Comment (nyalldawson) [inline src/app/vertextool/qgsvertextool.cpp:583]: It took me a while to understand this logic. Maybe a comment like:

 // If this feature is not visible, and we've already determined that we have to show users a warning about selecting
 // invisible vertices, then we don't need to check through the vertices for this feature. We'll be showing the warning anyway regardless of the outcome!

Comment (uclaros) [reply src/app/vertextool/qgsvertextool.cpp:585]: OK! Also renamed `invisibleVertexSelected' to `showInvisibleFeatureWarning`. I think the intention is clearer now!

**Comment by uclaros** [comment]
Any suggestion for the warning message wording as a native speaker?

File: src/app/vertextool/qgsvertextool.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/app/vertextool/qgsvertextool.cpp:585]: Much better, thanks!

**Comment by nyalldawson** [comment]
Hmm. maybe "One or more selected vertices are not currently visible on the map. These will not be modified."

**Comment by uclaros** [comment]
It's not that they will not be modified, it's actually that they will not be selected at all though...
I think the concept of *symbology classes* should be in the message somehow.

**Comment by roya0045** [comment]
@uclaros does that mean that if you selected vertices on a layer, zoom in a move them, they will be deselected or unmoved?

**Comment by uclaros** [comment]
@roya0045 no, the logic is simple and intuitive and is as follows:
When your symbology excludes some features, either because there is no matching class for it (eg delete the *all other values* class), or because a class's visibility is unchecked from the layers panel, then those features' vertices cannot be selected when dragging a rectangle with the vertex tool (just like they already cannot be selected by hover and click with the vertex tool). If however the drag rectangle contains (topologically) vertices from such features, a warning is issued stating that invisible vertices were not selected. Once the selection is done, everything works as before.

**Pull request merged**